[IMP] 7.0 add m2x options for view

azoellner opened the pull request:
Providing m2x widget options (`create`, `create_edit`, `m2o_dialog`, `limit`, `search_more`) at the view level via the `context` field of the `ir.actions.act_window`.

azoellner committed:
Reworked code for readability and consistent use of config option values

azoellner committed:
Added m2x options at the form level
via "context" of the "ir.actions.act_window"

azoellner committed:
Added documentation for the m2x options at the form level

OSguard commented:
azoellner is covered by the company CLA from initOS

hbrunn commented:
@OSguard then I suggest to check with @OCA/board to have this handle included, otherwise those messages will keep coming

hbrunn commented:
@azoellner I think if would be a lot more elegant to do your changes in https://github.com/OCA/web/pull/305/files#diff-d93aa186928a77371c865d1748af9b38R31, letting the appropriate context option win against the config parameter. Then the rest of the diff would only be you refactorings, which are badly needed i think

hbrunn commented:
Porting this to 8.0 at once would be highly appreciated, but it's not blocking

azoellner committed:
Fixed an undefined variable

pedrobaeza commented:
Do you think these changes should be under the same module? They are for action windows, not for m2x fields.

hbrunn commented:
@pedrobaeza to my understanding, this code allows to set the options via the context, I think it's fine to have it in the same module

azoellner commented:
>  letting the appropriate context option win against the config parameter

For me, the most reasonable priorities seem to be (from lowest to highest)
1. default, i.e. the implicit global (Odoo database) level
2. `ir_config_parameter`, i.e. the explicit global level
3. `context` of the action/view, i.e. the form view level
4. `options` of the field widget, i.e. the individual field level

as they get more and more specific. I think this reflects in the code, as the checks successively override the option value that is ultimately used.

hbrunn commented:
@azoellner agreed, and that's exactly what will happen if you follow my suggestion, just less messy as it is now

azoellner commented:
@hbrunn Now I understand your suggestion. Agreed, moving the code into that function will avoid the ugly code duplication.

hbrunn commented:
@azoellner setting to needs fixing until you find the time to move the code

pedrobaeza commented:
@azoellner, are you going to finish this?

azoellner committed:
Refactored out common code

azoellner committed:
Refactored getting the 'web_m2x_options'

simplified determining their values by moving the code
to a single, central point in method 'get_options'

azoellner commented:
Sorry it took me so long. Hopefully my code now looks a bit better. Refactored the option-related common code from both the 'many2one' and the 'many2manytags' widget into common functions, and then refactored the multiple checks for determining the option values to a single common location.

azoellner committed:
[pylint] Fixed pylint errors and warnings

pedrobaeza commented:
There's still 2 missing pylint problems:

```
web_m2x_options/__openerp__.py:3: [C8102(manifest-required-key), ] Missing required key "license" in manifest file
web_m2x_options/__openerp__.py:3: [C8106(manifest-version-format), ] Wrong Version Format "0.1" in manifest file. Regex to match: "(7.0)\.\d+\.\d+\.\d+"
```

azoellner commented:
> There's still 2 missing pylint problems

I didn't want to touch those, because I'm not sure about them. A reasonable `version` seems to be `7.0.0.1`, or maybe `7.0.0.2`. But `license` is not clear to me, since I'm not the original author. What do you suggest?

pedrobaeza commented:
You have to put 'AGPL-3' as license, and the version according OCA guidelines is 7.0.1.0.0 (https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers)

azoellner committed:
[pylint] Fixed more pylint errors and warnings

pedrobaeza reviewed (approved):
Waiting @hbrunn's approval
